### ISSUE TITLE
Edge bug with list-styles

### ISSUE THREAD

**Issue by ststimac** [webhintio/webhint.io#210]
I haven't checked Firefox, but I've checked Chrome, and this appears to be Edge specific. I'm not sure if there's something in the CSS causing this or if I need to file a bug against Edge & remove this issue.

When the home page first loads, and I hover over Documentation, there is extra spacing between the list items. 

![nav-first load](https://user-images.githubusercontent.com/18073131/30565694-29d406a2-9c7e-11e7-9112-c5e8d47b1d4c.PNG)

When I navigate to a subpage and then back to the home page, list items under Documentation are displaying properly with no extra padding/spacing. 

![nav-second-load](https://user-images.githubusercontent.com/18073131/30565706-32824494-9c7e-11e7-88f6-7b90ab9b502b.PNG)

If I refresh the home page though, that extra spacing appears again and doesn't go away until I navigate to a subpage and then back to the home page.

**Comment by ststimac** [comment]
This is also happening on subpages, if the user refreshes the page, the `list-style: none;` doesn't seem to be active, even though upon inspecting element it is. Unchecking and re-checking the style activates it. 

Checked Edge Portal and not seeing an issues with navigation there so continuing to investigate this on the sonar site in Edge only.

**Comment by molant** [comment]
Maybe something related to your build? Ask @gregwhitworth or @FremyCompany, maybe they know something about it?

**Comment by ststimac** [comment]
Going to update to the latest build and see if the issue persists.

**Comment by gregwhitworth** [comment]
If it still repros file the bug - possibly reduce it if you can. I'm seeing it on a ~3 week old build.

**Comment by FremyCompany** [comment]
That's the list-item gets hidden at the same time it gets its list-style removed. Since it gets hidden, it does not receive the notification that should be responsible for removing the ::marker pseudo-element. When it gets shown again, the element thinks its state is up to date and doesn't try to recompute marker state.
I thought we had considered this bug for fixing in RS3 but maybe I'm wrong...

**Comment by FremyCompany** [comment]
The bug is marked as resolved here: `Bug 104357: When showing a previously-hidden list-item with list-style-type none, the bullets are shown (if it had a ::marker prior to being hidden)`. @gregwhitworth could you figure out with Jacques why this still repros? Maybe we're looking at the wrong branch, or the fix didn't actually work.

**Comment by mixrich** [comment]
Hi. I had the same bug with Edge in my project.
I can't find any approvement that it's the Edge problem and will or was in solved or not by Microsoft, but I found solution for IE and Edge.

You can add style
`list-style-image: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkYAAAAAYAAjCB0C8AAAAASUVORK5CYII=);` for your list items. It's a 1x1 transparent PNG image, which is working well in all browsers and make lists looks without bullets. 

P.S. Sorry for my English

**Comment by FremyCompany** [comment]
@mixrich 

So, to be more precise, the issue was not fixed in latest release, and was actually made worse (as in: more frequent) by a regression in the page load algorithm. Because this problem became even more prominent in the latest release (Edge16) a decision has been taken to service a fix for this issue for that version as soon as we have a fix for it (currently, we still do not, but I'm personally tracking the status weekly).

I can confirm your workaround is going to work, but it's seriously suboptimal.

**Comment by Ygnis** [comment]
I found a solution which  works for me,
it will remove the space and make list-item markers transparent : 
ul{ padding: 0; margin:0; color: rgba(255, 255, 255, 0);}
Since the fourth number is 0, you could choose the color you want the result will be the same.
Sorry if I made some mistakes, English is not my first language.

**Comment by Ygnis** [comment]
I found a simple solution which works for me ; it will remove the space and make the list-item markers transparent: ul{ padding: 0; margin: 0; color: rgba(255, 255, 255, 0); }. The fourth number of rgba must be 0 for the complete  transparency of markers.
Sorry for my English.

**Comment by Acidic9** [comment]
I also have this issue, even going to developer tools and unchecking and rechecking the `list-item-type` property (disable then enable) makes the list style dots disappear...

**Comment by molant** [comment]
For those interested [here is a Twitter thread explaining a bit more the situation](https://twitter.com/fremycompany/status/950856311083683840)

**Comment by xpding** [comment]
@Ygnis 

Your solution remove the bullet in front of the list, but the line height still not right so there is gap around the list item. Not ideal but better than having the bullet and the gap...

**Comment by iDigress** [comment]
Been having the same issue, but with ul in navbar.  FremyCompany gave me an idea that worked well.  Basically he said it's because it's hiding them first then decides there's no need to provide any further formating.  I already had my list-style-type: none; before the display: none; so what I did was this: 

​#navBar ul ul {
    list-style-type: none;
}
#navBar ul ul {
    display: none;
    position: absolute;
}

Redundant? Yes. But it worked like a charm after banging my head against a wall for 4 hours on a 1 hour project before stumbling into this forum.  THANKS!!!

**Comment by FritzTGreen** [comment]
As the Twitter thread supplied by @molant explains, Edge is building the page layout before the CSS styles are loaded. Apparently, because the lists are hidden, Edge thinks it is done setting the state of the bullet markers and does not reconsider their state when the CSS styles are eventually applied. At that point, for reasons beyond my comprehension, the markers are already part of the layout and their state seemingly can't be altered. Even using javascript to add an inline style of list-style: none directly to the list does not work, according to my tests.

This likely explains why the problem goes away if you navigate to another page on the same site: the css files are cached and already available when Edge begins building/rendering the layout. Refresh the page, and the cached styles are ignored (unless you have Developer Tools open - I'm not sure why), so the problem comes back.

Since our sites use hidden lists for dropdown navigation menus, I have come up with a workaround that uses jQuery to clone each hidden list element, remove the original, and append the clone in its place when the page is loaded and ready. Since the reinserted clone is technically a new set of DOM elements, the result is that Edge is forced to rebuild and re-render it, correctly applying the now-available styles as it does. We only run the workaround if the browser is IE or Edge.

Here is a simplified form of our workaround:

`$(function(){
       if (document.documentMode || /Edge/.test(navigator.userAgent)) {
	     cloneDropDowns();
       }
}

/**
 * Fix for IE/Edge list-item=none render bug on list items hidden when page is initially loaded.
 */
function cloneNavDropDowns () {
	var menu_clone;
	var parent;
	var hidden_lists = $('.menu-dropdown');
	hidden_lists.each(function(){
		parent = $(this).parent();
		menu_clone = $(this).clone();
		$(this).remove();
		menu_clone.appendTo(parent);
	});
}`

In this example, the function acts on all lists with a class of ".menu-dropdown", i.e.: `<ul class=".menu-dropdown">`. You could simplify it further if you only have one hidden list to fix.

When Microsoft finally resolves the issue, this workaround will be obsolete, but until then, I hope this is helpful to someone.

Fritz

**Comment by anzolo** [comment]
I find nice work-around on https://stackoverflow.com/questions/20356311/internet-explorer-11-ignores-list-stylenone-on-the-first-load
It works for me.

**Comment by Ashish-Dwi** [comment]
I have found the solution for this issue with two steps.

1. For liist style problem use 
ul li{list-style-image: url(data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7);}

2. For spacing use
ul li{float:left; width:100%;}
ul li a{float:left; width:100%;}

**Comment by WolfieZero** [comment]
I had this exact issue and tried doing the basic fixes suggested with CSS to no avail. @FritzTGreen's JS solution did work for me perfectly.

**Comment by MaartenW** [comment]
Thanks @FritzTGreen! I made your solution work on any and all hidden lists. Doing a detach and append works as well.

    // Mitigate IE/Edge bug showing bullets on lists which are hidden when loading the page
    $(document).ready(function(){
        if (document.documentMode || /Edge/.test(navigator.userAgent)) {
            $('ul:hidden').each(function(){
                $(this).parent().append($(this).detach());
            });
        }
    });

**Comment by FritzTGreen** [comment]
@MaartenW That is a clean, beautiful piece of code. I had tried to do the clone and append in the same line, but it wouldn't work. I didn't think of using detach. Thanks!

**Comment by cmckinnonfb** [comment]
@Ashish-Dwi, thank you for your solution, it worked.

Its 2018, I CANNOT believe we still have to jump through hoops like this to deal with Microsoft ineptitude.

**Comment by FritzTGreen** [comment]
I'm surprised that the CSS fixes work for some people and not others (like me). I would be curious to know what is different.
I am not surprised that we still have to jump through Microsoft hoops, though.

**Comment by mikelnorth** [comment]
I have the same problem in my application, but it is not always reproduced. When does the problem become popular enough that Microsoft fixes it instead of us having to do hacks around to support them?

**Comment by ststimac** [comment]
I believe this has been fixed in one of the recent Windows Insider builds. I'm no longer seeing this issue on the two sites where I was seeing it.

**Comment by cmckinnonfb** [comment]
@ststimac, thank you for the update.  Hopefully we will see the fix out in the wild soon.

**Comment by molant** [comment]
Closing this issue as it's fixed in the latest stable version of Edge.

**Comment by nathanhannig** [comment]
can you provide more details on the alpha version of Edge, a link, or how we can check it out?

**Comment by gregwhitworth** [comment]
@nathanhannig It's with RS5 which began shipping on November 13th. You can either update your box or grab a VM from [here](https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/).

**Comment by anthonyekosky** [comment]
> Thanks @FritzTGreen! I made your solution work on any and all hidden lists. Doing a detach and append works as well.
> 
> ```
> // Mitigate IE/Edge bug showing bullets on lists which are hidden when loading the page
> $(document).ready(function(){
>     if (document.documentMode || /Edge/.test(navigator.userAgent)) {
>         $('ul:hidden').each(function(){
>             $(this).parent().append($(this).detach());
>         });
>     }
> });
> ```
Others didn't work for me. This worked perfectly. Thank you!

**Comment by ayushkov** [comment]
If there are nested lists
```// Mitigate IE/Edge bug showing bullets on lists which are hidden when loading the page
$(document).ready(function(){
    if (document.documentMode || /Edge/.test(navigator.userAgent)) {
        $('ul:hidden').each(function(){
			$(this).replaceWith($(this));
        });
    }
});
```

**Comment by danielbarrosai** [comment]
To fix this bug, just write the following code:

`ul li {
list-style:none;
list-style-image:url(data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7);
}`

Hope this helps. :)

**Comment by shashwattiwari** [comment]
> I have found the solution for this issue with two steps.
> 
> 1. For liist style problem use
>    ul li{list-style-image: url(data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7);}
> 2. For spacing use
>    ul li{float:left; width:100%;}
>    ul li a{float:left; width:100%;}

Thanks Ashish.. it worked.
